### DISCUSSION TITLE
Make it possible to (easily) figure out if a configuration can map between types

### DISCUSSION THREAD

**Discussion by erik-kallen** [LuckyPennySoftware/AutoMapper#3917] [NA]
It would be nice with a method `IConfigurationProvider.CanMap(Type sourceType, Type destinationType)`.

I want this because I have tests that validate that a mapper will be able to perform all mappings I will be asking it to do.

**Comment by lbargaoanu** [comment]
https://docs.automapper.org/en/latest/Configuration-validation.html
See also [this](https://github.com/AutoMapper/AutoMapper/blob/f0635d27bfff8be5ff1831a5b7e1717757efe4e4/src/AutoMapper/Configuration/MapperConfiguration.cs#L43) or [this](https://github.com/AutoMapper/AutoMapper/blob/f0635d27bfff8be5ff1831a5b7e1717757efe4e4/src/AutoMapper/Internal/InternalApi.cs#L103).

**Reply by erik-kallen** [reply]
Thank you for your quick reply. I don't think any of those work for me, though.

1. The link to the documentation only talks about validating existing mappings, I want to validate that a mapping exists.
2. `BuildExecutionPlan` will happily (it seems) return a plan for the mapping whether or not it exists, and
3. The third link, `AssertConfigurationIsValid(TypeMap)` might work, but I would need to construct my own TypeMap to pass in to it, which doesn't seem right to me.

To be clear, what I want to do is something like

```
class MyDateTimeModel {}

[Test]
public void AssertMappingExists(IMapper mapper)
{
    Assert.That(mapper.CanMap<DateTime, MyDateTimeModel>());
}
```

What I am looking for can be done approximately with

```
mapper.ConfigurationProvider.Internal().GetAllTypeMaps().Any(t => t.SourceType == sourceType && t.DestinationType == destinationType)
```

but that doesn't work fully when (for example) the types are enumerables of some inner type (since AutoMapper automatically creates a mapping from `IEnumerable<TSource>` to `IEnumerable<TDest>` as long as the `TSource` => `TDest` mapping exists)

**Reply by lbargaoanu** [reply]
I don't see why you need anything other than validating the configuration. If your configuration is valid, a mapping for the types there will probably work. But nothing really ensures that other than an unit test. So what's _really_ the use for a more specific question? But whatever it might be, feel free to write a function that does that. I've already pointed you in the right direction.

**Reply by erik-kallen** [reply]
I have written a function that does it (minus the enumerable thing because I don't want to replicate all logic automapper has for creating implicit mappings).

I can't understand, though why you "don't see why you need anything other than validating the configuration". You are correct that I only need to "validate the configuration", but validating the configuration means that I need to validate that certain mappings exist , not just validating the mappings that do exist.

As I said, this is because I want to write a test that can ensure that I won't get exceptions about missing mappings when I'm in production.

**Reply by lbargaoanu** [reply]
You're describing exactly what configuration validation already does for you, free of charge :)
What's not covered is, for example, when you're trying to map some types that are not anywhere in your mappings, some root types are missing. But surely that's an obvious error :)

**Reply by erik-kallen** [reply]
And that "obvious" error is what I want to catch in my tests.

**Reply by lbargaoanu** [reply]
Just call `Map<Destination>(new Source())`. But you're way off track here. For anything other than the most trivial map, you need a real unit test that verifies what you wrote in the mapping. _That_'s what you really need to test, your own code.

**Reply by erik-kallen** [reply]
The suggestion works, though `Map<TSource, TDestination>(default(Source))` works better in case there is no public constructor

**Reply by lbargaoanu** [reply]
Mapping from `null` will always succeed.
